Feature/select implementation utils

lucas-baran opened the pull request:
Most of these scripts come from the UnityDataValidation package and Corsairs, I updated them to fit more general needs.

# Changelog

## [1.2.0] - 2024-04-24
### Added
- Add `ReflectionUtilities` static class
  - Get implementation types from base type
  - Create instances of implementation types from base type
- Add `SelectImplementationDropdown` to select the implementation of a base type in the inspector
- Add property context menu utilities scripts:
  - Add `IPropertyContextMenuCallback` and `PropertyContextMenuCallbackFetcher` to make it easier to add custom property context menus
  - Add `CreateAssetContextMenu` to quickly create scriptable objects by selecting an implementation
![create_asset_context_menu_0](https://github.com/FishingCactus/UnityCommonCode/assets/61627716/62c41483-2f4e-4530-80be-2531d7401dc0)
  - Add `VectorContextMenu` to quickly set values of vectors
![vector_context_menu_0](https://github.com/FishingCactus/UnityCommonCode/assets/61627716/40c3593a-5db1-41f9-a1b0-15f56287480e)
- Add `HierarchyScriptDropHandler` which allows people to drag and drop scripts inside the scene hierarchy to directly create an empty game object with the script attached
![hierarchy_drop_script_0](https://github.com/FishingCactus/UnityCommonCode/assets/61627716/88c14652-c832-4be2-868b-afc5cfc3c50d)
- Add `ReorderableSubAssetList` which allows to create sub assets in a list by selecting the implementation
![reordable_sub_asset_list_0](https://github.com/FishingCactus/UnityCommonCode/assets/61627716/dea182ca-9bb0-4a3e-bf0d-a9eae7f7b224)
- Add `HierarchyUtilities` with utility functions to rename and create game objects in the scene hierarchy
- Add utility functions in `InspectorUtilities` to get the type of a serialized property that works for lists and arrays

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110)
<!-- Reviewable:end -->

lucas-baran committed:
add ReflectionUtilities

- moved from UnityDataValidation package

lucas-baran committed:
add TypeDropdownItem

- moved from UnityDataValidation package

lucas-baran committed:
add SelectImplementationDropdown

- moved from UnityDataValidation package

lucas-baran committed:
add IEnableable interface

- moved from UnityDataValidation package

lucas-baran committed:
add ReorderableSubAssetList

- moved from UnityDataValidation package

lucas-baran committed:
EditorUtils: add GetPropertyType and TryGetArrayOrListElementType

lucas-baran committed:
EditorUtils: cs spacing

lucas-baran committed:
SelectImplementationDropdown: add non-generic implementation

lucas-baran committed:
generic SelectImplementationDropdown is not sealed

lucas-baran committed:
add IPropertyContextMenuCallback interface

lucas-baran committed:
add PropertyContextMenuCallbackFetcher

lucas-baran committed:
add CreateAssetContextMenu

lucas-baran committed:
add VectorContextMenu

lucas-baran committed:
add HierarchyUtilities

lucas-baran committed:
add HierarchyScriptDropHandler

lucas-baran committed:
CreateAssetContextMenu: CreateAssetOfType is static

lucas-baran committed:
TypeDropdownItem: add optional name

lucas-baran committed:
SelectImplementationDropdown: add optional name delegate + add category items

lucas-baran committed:
Revert "SelectImplementationDropdown: add optional name delegate + add category items"

This reverts commit 843cbb5a4e962c4f5f494373b3164c9112d1b50f.

lucas-baran committed:
CreateAssetContextMenu: add min dropdown width

lucas-baran committed:
CreateAssetContextMenu: add GetCreateAssetMenuAttributeInfo

lucas-baran committed:
package.json: update version number

lucas-baran committed:
udpate changelog

lucas-baran committed:
Revert "TypeDropdownItem: add optional name"

This reverts commit a4e20000dbd5f74e29e6504dd98e39d26e710a9a.

lucas-baran committed:
udpate changelog

fishingcactus-nathan-istace commented:
*[`Editor/EditorUtils.cs` line 67 at r1](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110#-NwEIyWE9jmrrBhOo_Ly:-NwEIyWE9jmrrBhOo_Lz:bz4i1re) ([raw file](https://github.com/FishingCactus/UnityCommonCode/blob/ef4587e27035448c31d9ce16cc8eb88262b4eb53/Editor/EditorUtils.cs#L67)):*
> ```Smalltalk
>         }
> 
>         public static bool TryGetArrayOrListElementType(
> ```

This should follow the new coding standards (Denied: one parameter per line)


<!-- Sent from Reviewable.io -->

fishingcactus-nathan-istace commented:
*[`Editor/EditorUtils.cs` line 80 at r1](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110#-NwEKCdR10aXP6x14OaM:-NwEKCdR10aXP6x14OaN:b-7ndqnd) ([raw file](https://github.com/FishingCactus/UnityCommonCode/blob/ef4587e27035448c31d9ce16cc8eb88262b4eb53/Editor/EditorUtils.cs#L80)):*
> ```Smalltalk
> 
>             if( list_type.IsGenericType
>                 && list_type.GetGenericTypeDefinition() == typeof( List<> )
> ```

Could be generalized to handle typeof( IReadOnlyList<> )


<!-- Sent from Reviewable.io -->

fishingcactus-nathan-istace commented:
*[`Editor/ContextMenus/PropertyContextMenus/CreateAssetContextMenu.cs` line 20 at r1](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110#-NwEL0mK3Nlram9yFieG:-NwEL0mK3Nlram9yFieH:b-gdt0u5) ([raw file](https://github.com/FishingCactus/UnityCommonCode/blob/ef4587e27035448c31d9ce16cc8eb88262b4eb53/Editor/ContextMenus/PropertyContextMenus/CreateAssetContextMenu.cs#L20)):*
> ```Smalltalk
> 
>         public void OnPropertyContextMenu(
>             Rect context_menu_rect,
> ```

CS (Denied: One parameter per line)


<!-- Sent from Reviewable.io -->

Skeloson reviewed (commented):
> Reviewed 32 of 33 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110)* status: 32 of 33 files reviewed, 4 unresolved discussions (waiting on @fishingcactus-nathan-istace, @LeanderSimonart, @LouisNillusFC, @lucas-baran, @MaxDaniels, @PetitT, @rascar-capac, @TheEmidee, and @ThibNach)

___
*[`Runtime/Helpers/ReflectionUtilities.cs` line 15 at r1](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110#-NwEPK9P4Qah-dmbMOgU:-NwEPK9P4Qah-dmbMOgV:b-2pu1ko) ([raw file](https://github.com/FishingCactus/UnityCommonCode/blob/ef4587e27035448c31d9ce16cc8eb88262b4eb53/Runtime/Helpers/ReflectionUtilities.cs#L15)):*
> ```Smalltalk
>             )
>         {
>             return type.IsClass && type.IsSealed && type.IsAbstract;
> ```

A static class should be Sealed and Abstract ?
___
*[`Runtime/Helpers/ReflectionUtilities.cs` line 45 at r1](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110#-NwEQ02k2qXVxy_0GI7j:-NwEQ02k2qXVxy_0GI7k:b-odlo3v) ([raw file](https://github.com/FishingCactus/UnityCommonCode/blob/ef4587e27035448c31d9ce16cc8eb88262b4eb53/Runtime/Helpers/ReflectionUtilities.cs#L45)):*
> ```Smalltalk
>             CheckImplementationDelegate( ref implementation_delegate );
> 
>             bool CanActivateImplementation(
> ```

Do you need this methods since you use it only one ? Cant you use a local method here ?


<!-- Sent from Reviewable.io -->

lucas-baran reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110)* status: 32 of 33 files reviewed, 4 unresolved discussions (waiting on @fishingcactus-nathan-istace, @LeanderSimonart, @LouisNillusFC, @MaxDaniels, @PetitT, @rascar-capac, @Skeloson, @TheEmidee, and @ThibNach)

___
*[`Editor/EditorUtils.cs` line 67 at r1](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110#-NwEIyWE9jmrrBhOo_Ly:-NwEtRYC5DXPuqN5fzpa:bdyicvh) ([raw file](https://github.com/FishingCactus/UnityCommonCode/blob/ef4587e27035448c31d9ce16cc8eb88262b4eb53/Editor/EditorUtils.cs#L67)):*
<details><summary><i>Previously, fishingcactus-nathan-istace (NathanIstace) wrote…</i></summary><blockquote>

This should follow the new coding standards (Denied: one parameter per line)
</blockquote></details>

Not required in the coding standards, though I will move the functions I added in a separate class because they are not related to this class thus shouldn't be in there
___
*[`Editor/EditorUtils.cs` line 80 at r1](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110#-NwEKCdR10aXP6x14OaM:-NwEthGj90PeVm9Hbxej:bjsv1l2) ([raw file](https://github.com/FishingCactus/UnityCommonCode/blob/ef4587e27035448c31d9ce16cc8eb88262b4eb53/Editor/EditorUtils.cs#L80)):*
<details><summary><i>Previously, fishingcactus-nathan-istace (NathanIstace) wrote…</i></summary><blockquote>

Could be generalized to handle typeof( IReadOnlyList<> )
</blockquote></details>

I don't think it would work or I would have to check if the property type can be assigned to the IReadOnlyList<> type, but since it's meant to be used for get the element type of a SerializedProperty (which is either a List or an Array) I don't think it would all that useful to do that
___
*[`Editor/ContextMenus/PropertyContextMenus/CreateAssetContextMenu.cs` line 20 at r1](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110#-NwEL0mK3Nlram9yFieG:-NwEtP2BCJKR13aiJaZf:bi1oz4r) ([raw file](https://github.com/FishingCactus/UnityCommonCode/blob/ef4587e27035448c31d9ce16cc8eb88262b4eb53/Editor/ContextMenus/PropertyContextMenus/CreateAssetContextMenu.cs#L20)):*
<details><summary><i>Previously, fishingcactus-nathan-istace (NathanIstace) wrote…</i></summary><blockquote>

CS (Denied: One parameter per line)
</blockquote></details>

Not required in the coding standards
___
*[`Runtime/Helpers/ReflectionUtilities.cs` line 15 at r1](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110#-NwEPK9P4Qah-dmbMOgU:-NwEsd2iEntLmUmYkS_t:b-bt2e5e) ([raw file](https://github.com/FishingCactus/UnityCommonCode/blob/ef4587e27035448c31d9ce16cc8eb88262b4eb53/Runtime/Helpers/ReflectionUtilities.cs#L15)):*
<details><summary><i>Previously, Skeloson (Jérémy Leclef) wrote…</i></summary><blockquote>

A static class should be Sealed and Abstract ?
</blockquote></details>

I know it's strange but that's what a static class is under the hood, which makes sense if you think about it: abstract prevents you from creating instances of that class and sealed prevents you from inheriting from it
___
*[`Runtime/Helpers/ReflectionUtilities.cs` line 45 at r1](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110#-NwEQ02k2qXVxy_0GI7j:-NwEt7hf5y6lYKjXshGb:bgwnqeu) ([raw file](https://github.com/FishingCactus/UnityCommonCode/blob/ef4587e27035448c31d9ce16cc8eb88262b4eb53/Runtime/Helpers/ReflectionUtilities.cs#L45)):*
<details><summary><i>Previously, Skeloson (Jérémy Leclef) wrote…</i></summary><blockquote>

Do you need this methods since you use it only one ? Cant you use a local method here ?
</blockquote></details>

I can move it to be not local if that's what you meant, that way I wouldn't be capturing parameters either


<!-- Sent from Reviewable.io -->

Skeloson reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110)* status: 32 of 33 files reviewed, 3 unresolved discussions (waiting on @fishingcactus-nathan-istace, @LeanderSimonart, @LouisNillusFC, @lucas-baran, @MaxDaniels, @PetitT, @rascar-capac, @TheEmidee, and @ThibNach)

___
*[`Runtime/Helpers/ReflectionUtilities.cs` line 45 at r1](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110#-NwEQ02k2qXVxy_0GI7j:-NwFD76w3EVZ2LJpmkv3:b-fynghy) ([raw file](https://github.com/FishingCactus/UnityCommonCode/blob/ef4587e27035448c31d9ce16cc8eb88262b4eb53/Runtime/Helpers/ReflectionUtilities.cs#L45)):*
<details><summary><i>Previously, lucas-baran (Lucas Baran) wrote…</i></summary><blockquote>

I can move it to be not local if that's what you meant, that way I wouldn't be capturing parameters either
</blockquote></details>

It more that I dont get the profit of a method inside a method in this case ^^'


<!-- Sent from Reviewable.io -->

TheEmidee reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 33 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110#-:-NwJYW-a3OXPq6m318oK:bnfp4nl)* status: all files reviewed, 3 unresolved discussions (waiting on @fishingcactus-nathan-istace, @LeanderSimonart, @LouisNillusFC, @lucas-baran, @MaxDaniels, @PetitT, @rascar-capac, and @ThibNach)



<!-- Sent from Reviewable.io -->

MaxDaniels reviewed (changes requested):
> Reviewed 32 of 33 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110)* status: all files reviewed, 4 unresolved discussions (waiting on @fishingcactus-nathan-istace, @LeanderSimonart, @LouisNillusFC, @lucas-baran, @PetitT, @rascar-capac, and @ThibNach)

___
*[`Editor/ContextMenus/PropertyContextMenus/VectorContextMenu.cs` line 51 at r1](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110#-NwJlJU35XhYvpjnu9MX:-NwJlJU35XhYvpjnu9MY:bi6lqja) ([raw file](https://github.com/FishingCactus/UnityCommonCode/blob/ef4587e27035448c31d9ce16cc8eb88262b4eb53/Editor/ContextMenus/PropertyContextMenus/VectorContextMenu.cs#L51)):*
> ```Smalltalk
>             {
>                 case SerializedPropertyType.Vector2:
>                     property.vector2Value = new Vector2( x, y );
> ```

Waiting feedbacks in CodingStandard slack channel before validating cases without {}.


<!-- Sent from Reviewable.io -->

lucas-baran committed:
move GetPropertyType function from EditorUtils to InspectorUtilities

lucas-baran committed:
update changelog

lucas-baran reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110)* status: 31 of 35 files reviewed, 4 unresolved discussions (waiting on @fishingcactus-nathan-istace, @LeanderSimonart, @LouisNillusFC, @MaxDaniels, @PetitT, @rascar-capac, @Skeloson, @TheEmidee, and @ThibNach)

___
*[`Editor/EditorUtils.cs` line 67 at r1](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110#-NwEIyWE9jmrrBhOo_Ly:-NwK8niK0yC806Vr5TXl:bxkjgi6) ([raw file](https://github.com/FishingCactus/UnityCommonCode/blob/ef4587e27035448c31d9ce16cc8eb88262b4eb53/Editor/EditorUtils.cs#L67)):*
<details><summary><i>Previously, lucas-baran (Lucas Baran) wrote…</i></summary><blockquote>

Not required in the coding standards, though I will move the functions I added in a separate class because they are not related to this class thus shouldn't be in there
</blockquote></details>

Moved it inside `InspectorUtilities`
___
*[`Runtime/Helpers/ReflectionUtilities.cs` line 45 at r1](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110#-NwEQ02k2qXVxy_0GI7j:-NwJBoR84nT_IRPJF6Ew:bsuclsi) ([raw file](https://github.com/FishingCactus/UnityCommonCode/blob/ef4587e27035448c31d9ce16cc8eb88262b4eb53/Runtime/Helpers/ReflectionUtilities.cs#L45)):*
<details><summary><i>Previously, Skeloson (Jérémy Leclef) wrote…</i></summary><blockquote>

It more that I dont get the profit of a method inside a method in this case ^^'
</blockquote></details>

So I tried to change it and I remember why I did it this way no, I'm using the function in the `GetImplementations` function which is a `Func<Type, bool>` but I also need the `implementation_delegate` so it allows me to capture it inside the local function which wouldn't be possible if it weren't local.
I could as well define a function with the conditions required for creating instances and use a lambda to combine `CanActivateImplementation` and `implementation_delegate` if you prefer


<!-- Sent from Reviewable.io -->

TheEmidee reviewed (commented):
> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110)* status: 31 of 35 files reviewed, 4 unresolved discussions (waiting on @fishingcactus-nathan-istace, @LeanderSimonart, @LouisNillusFC, @MaxDaniels, @PetitT, @rascar-capac, @Skeloson, and @ThibNach)

___
*[`Editor/ContextMenus/PropertyContextMenus/VectorContextMenu.cs` line 51 at r1](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110#-NwJlJU35XhYvpjnu9MX:-NwKAaOB66BtQlKP0RQV:b3u57wz) ([raw file](https://github.com/FishingCactus/UnityCommonCode/blob/ef4587e27035448c31d9ce16cc8eb88262b4eb53/Editor/ContextMenus/PropertyContextMenus/VectorContextMenu.cs#L51)):*
<details><summary><i>Previously, MaxDaniels wrote…</i></summary><blockquote>

Waiting feedbacks in CodingStandard slack channel before validating cases without {}.
</blockquote></details>

should be mandatory even for `case` statements


<!-- Sent from Reviewable.io -->

lucas-baran committed:
udpate .editorconfig

lucas-baran committed:
VectorContextMenu: add brackets for switch-cases

lucas-baran committed:
update changelog

lucas-baran reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110)* status: 29 of 36 files reviewed, 4 unresolved discussions (waiting on @fishingcactus-nathan-istace, @LeanderSimonart, @LouisNillusFC, @MaxDaniels, @PetitT, @rascar-capac, @Skeloson, @TheEmidee, and @ThibNach)

___
*[`Editor/ContextMenus/PropertyContextMenus/VectorContextMenu.cs` line 51 at r1](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110#-NwJlJU35XhYvpjnu9MX:-NwQA_F1D_XuaARbM26-:b-896fix) ([raw file](https://github.com/FishingCactus/UnityCommonCode/blob/ef4587e27035448c31d9ce16cc8eb88262b4eb53/Editor/ContextMenus/PropertyContextMenus/VectorContextMenu.cs#L51)):*
<details><summary><i>Previously, TheEmidee (Michael Delva) wrote…</i></summary><blockquote>

should be mandatory even for `case` statements
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

LouisNillusFC reviewed (approved):
> Reviewed 3 of 4 files at r2, 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110)* status: 35 of 36 files reviewed, 4 unresolved discussions (waiting on @fishingcactus-nathan-istace, @LeanderSimonart, @MaxDaniels, @PetitT, @rascar-capac, @Skeloson, and @ThibNach)



<!-- Sent from Reviewable.io -->

LeanderSimonart reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110#-:-Nwhn3Hz-YVFEGYibfMy:bnfp4nl)* status: all files reviewed, 4 unresolved discussions (waiting on @fishingcactus-nathan-istace, @MaxDaniels, @PetitT, @rascar-capac, @Skeloson, and @ThibNach)



<!-- Sent from Reviewable.io -->

rascar-capac reviewed (changes requested):
> Reviewed 29 of 33 files at r1, 4 of 4 files at r2, 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110)* status: all files reviewed, 9 unresolved discussions (waiting on @fishingcactus-nathan-istace, @lucas-baran, @MaxDaniels, @PetitT, @Skeloson, and @ThibNach)

___
*[`Editor/ContextMenus/PropertyContextMenus/CreateAssetContextMenu.cs` line 124 at r4](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110#-NwivrluCLaw51oZ39xq:-NwivrluCLaw51oZ39xr:b-zbss46) ([raw file](https://github.com/FishingCactus/UnityCommonCode/blob/2b8645e3b248e77bc4eb423722d2f6fcb4bf2152/Editor/ContextMenus/PropertyContextMenus/CreateAssetContextMenu.cs#L124)):*
> ```Smalltalk
>             ScriptableObject asset = ScriptableObject.CreateInstance( asset_type );
>             string asset_path = ( string )parameters[ 0 ];
>             string asset_name = $"SO_{asset_type.Name}";
> ```

the pattern could be a const
___
*[`Editor/ContextMenus/PropertyContextMenus/VectorContextMenu.cs` line 23 at r4](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110#-Nwj-Q0652ssmdT-6u9w:-Nwj-Q0652ssmdT-6u9x:b-vpbue0) ([raw file](https://github.com/FishingCactus/UnityCommonCode/blob/2b8645e3b248e77bc4eb423722d2f6fcb4bf2152/Editor/ContextMenus/PropertyContextMenus/VectorContextMenu.cs#L23)):*
> ```Smalltalk
>         }
> 
>         private static bool IsVectorProperty(
> ```

are float3 and friends serializable? if they are you could add them here
___
*[`Editor/DropDowns/Items/TypeDropdownItem.cs` line 9 at r4](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110#-Nwj-wlP5Ws2T-lCUXYK:-Nwj-wlP5Ws2T-lCUXYL:bb78zq7) ([raw file](https://github.com/FishingCactus/UnityCommonCode/blob/2b8645e3b248e77bc4eb423722d2f6fcb4bf2152/Editor/DropDowns/Items/TypeDropdownItem.cs#L9)):*
> ```Smalltalk
>     public sealed class TypeDropdownItem : AdvancedDropdownItem
>     {
>         public Type Type { get; }
> ```

missing initialization
___
*[`Editor/Utilities/InspectorUtilities.cs` line 62 at r4](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110#-Nwj6Ouf2nQJbXfW7Yj_:-Nwj6Ouf2nQJbXfW7Yja:b-jpiebf) ([raw file](https://github.com/FishingCactus/UnityCommonCode/blob/2b8645e3b248e77bc4eb423722d2f6fcb4bf2152/Editor/Utilities/InspectorUtilities.cs#L62)):*
> ```Smalltalk
>             FieldInfo found_field = null;
> 
>             for( Type current_type = property_type; found_field == null && current_type != null; current_type = current_type.BaseType )
> ```

you don't like `while` do you? :P
___
*[`Runtime/Helpers/ReflectionUtilities.cs` line 34 at r4](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110#-Nwj59aK7iPsBT2cWo9C:-Nwj1NQWEhhIyagrtnq9:bgof2b8) ([raw file](https://github.com/FishingCactus/UnityCommonCode/blob/2b8645e3b248e77bc4eb423722d2f6fcb4bf2152/Runtime/Helpers/ReflectionUtilities.cs#L34)):*
> ```Smalltalk
>             return AppDomain.CurrentDomain.GetAssemblies()
>                 .SelectMany( assembly => assembly.GetTypes() )
>                 .Where( type => base_type.IsAssignableFrom( type ) && implementation_delegate( type ) );
> ```

I would find an `ImplementationIsValid()` that returns `implementation_delegate ? implementation_delegate( type ) : true` more straightforward


<!-- Sent from Reviewable.io -->

fishingcactus-nathan-istace commented:
*[`Editor/DropDowns/Items/TypeDropdownItem.cs` line 9 at r4](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110#-Nwj-wlP5Ws2T-lCUXYK:-NwsA99Q3aHcFXBoiQ-l:bbsna2m) ([raw file](https://github.com/FishingCactus/UnityCommonCode/blob/2b8645e3b248e77bc4eb423722d2f6fcb4bf2152/Editor/DropDowns/Items/TypeDropdownItem.cs#L9)):*
<details><summary><i>Previously, rascar-capac (Théo Constant) wrote…</i></summary><blockquote>

missing initialization
</blockquote></details>

Actually I think it makes it clearer that it can only be initialized in the constructor, and I don't assume that Type ever has whatever default value would be assigned here.


<!-- Sent from Reviewable.io -->

fishingcactus-nathan-istace commented:
*[`Editor/EditorUtils.cs` line 80 at r1](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110#-NwEKCdR10aXP6x14OaM:-NwsCKHeCcoJbybNS6z0:b465rw4) ([raw file](https://github.com/FishingCactus/UnityCommonCode/blob/ef4587e27035448c31d9ce16cc8eb88262b4eb53/Editor/EditorUtils.cs#L80)):*
<details><summary><i>Previously, lucas-baran (Lucas Baran) wrote…</i></summary><blockquote>

I don't think it would work or I would have to check if the property type can be assigned to the IReadOnlyList<> type, but since it's meant to be used for get the element type of a SerializedProperty (which is either a List or an Array) I don't think it would all that useful to do that
</blockquote></details>

Anything that is a type of List<T> is also a type of IReadOnlyList<T>, right? We could have to implement a class that implements IReadOnlyList but does not inherit from List, and this function wouldn't work on that other class. I just think it removes usecases that could easily be handled.


<!-- Sent from Reviewable.io -->

fishingcactus-nathan-istace reviewed (approved):
> Reviewed 32 of 33 files at r1, 4 of 4 files at r2, 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110)* status: all files reviewed, 6 unresolved discussions (waiting on @lucas-baran, @MaxDaniels, @PetitT, @rascar-capac, @Skeloson, and @ThibNach)



<!-- Sent from Reviewable.io -->

lucas-baran committed:
add reference to Unity.Mathematics

lucas-baran committed:
VectorContextMenu: add context menu for more vector types

Additional types:
- byte
- short
- uint
- int
- long
- float
- double
- float2
- float3
- float4
- int2
- int3
- int4

lucas-baran committed:
ReflectionUtilities: add IsAssignableToGenericType function

lucas-baran committed:
ReflectionUtilities: add TryGetEnumerableElementType function

lucas-baran committed:
InspectorUtilities: use ReflectionUtilities.TryGetEnumerableElementType

lucas-baran committed:
InspectorUtilities: remove TryGetArrayOrListElementType

lucas-baran committed:
move InspectorUtilities inside FishingCactus namespace

lucas-baran reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110)* status: all files reviewed, 6 unresolved discussions (waiting on @MaxDaniels, @PetitT, @rascar-capac, @Skeloson, and @ThibNach)

___
*[`Editor/EditorUtils.cs` line 80 at r1](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110#-NwEKCdR10aXP6x14OaM:-NwsJ0a36UzGlqNeIvml:byv40rf) ([raw file](https://github.com/FishingCactus/UnityCommonCode/blob/ef4587e27035448c31d9ce16cc8eb88262b4eb53/Editor/EditorUtils.cs#L80)):*
<details><summary><i>Previously, fishingcactus-nathan-istace (NathanIstace) wrote…</i></summary><blockquote>

Anything that is a type of List<T> is also a type of IReadOnlyList<T>, right? We could have to implement a class that implements IReadOnlyList but does not inherit from List, and this function wouldn't work on that other class. I just think it removes usecases that could easily be handled.
</blockquote></details>

In that case I can check it with `IEnumerable<>` instead, also I moved it inside `ReflectionUtilities` because it's no longer Inspector-only related
___
*[`Editor/ContextMenus/PropertyContextMenus/CreateAssetContextMenu.cs` line 124 at r4](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110#-NwivrluCLaw51oZ39xq:-NwsSt2675kQuWpmyTej:bgaahvs) ([raw file](https://github.com/FishingCactus/UnityCommonCode/blob/2b8645e3b248e77bc4eb423722d2f6fcb4bf2152/Editor/ContextMenus/PropertyContextMenus/CreateAssetContextMenu.cs#L124)):*
<details><summary><i>Previously, rascar-capac (Théo Constant) wrote…</i></summary><blockquote>

the pattern could be a const
</blockquote></details>

Actually it could/should be a setting I think so that it follows the naming conventions of the project, tbh I won't be doing that now xD
___
*[`Editor/ContextMenus/PropertyContextMenus/VectorContextMenu.cs` line 23 at r4](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110#-Nwj-Q0652ssmdT-6u9w:-NwjdfSx9tIO6HEJZqPS:bx0h0rm) ([raw file](https://github.com/FishingCactus/UnityCommonCode/blob/2b8645e3b248e77bc4eb423722d2f6fcb4bf2152/Editor/ContextMenus/PropertyContextMenus/VectorContextMenu.cs#L23)):*
<details><summary><i>Previously, rascar-capac (Théo Constant) wrote…</i></summary><blockquote>

are float3 and friends serializable? if they are you could add them here
</blockquote></details>

Yes they are. Also added it for integer types and floating point types
Done.
___
*[`Editor/DropDowns/Items/TypeDropdownItem.cs` line 9 at r4](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110#-Nwj-wlP5Ws2T-lCUXYK:-NwsIK1E2Y6DXwBdnYgT:b-9jyvdn) ([raw file](https://github.com/FishingCactus/UnityCommonCode/blob/2b8645e3b248e77bc4eb423722d2f6fcb4bf2152/Editor/DropDowns/Items/TypeDropdownItem.cs#L9)):*
<details><summary><i>Previously, fishingcactus-nathan-istace (NathanIstace) wrote…</i></summary><blockquote>

Actually I think it makes it clearer that it can only be initialized in the constructor, and I don't assume that Type ever has whatever default value would be assigned here.
</blockquote></details>

What Nathan said, and we don't have a standards on this
___
*[`Editor/Utilities/InspectorUtilities.cs` line 62 at r4](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110#-Nwj6Ouf2nQJbXfW7Yj_:-NwsTARs4k5Iv0HngY03:b-492wau) ([raw file](https://github.com/FishingCactus/UnityCommonCode/blob/2b8645e3b248e77bc4eb423722d2f6fcb4bf2152/Editor/Utilities/InspectorUtilities.cs#L62)):*
<details><summary><i>Previously, rascar-capac (Théo Constant) wrote…</i></summary><blockquote>

you don't like `while` do you? :P
</blockquote></details>

Yeeah it's the same anyways right?
___
*[`Runtime/Helpers/ReflectionUtilities.cs` line 34 at r4](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110#-Nwj59aK7iPsBT2cWo9C:-NwsIlWgBjCw8FmV2d7Y:b-67gj92) ([raw file](https://github.com/FishingCactus/UnityCommonCode/blob/2b8645e3b248e77bc4eb423722d2f6fcb4bf2152/Runtime/Helpers/ReflectionUtilities.cs#L34)):*
<details><summary><i>Previously, rascar-capac (Théo Constant) wrote…</i></summary><blockquote>

I would find an `ImplementationIsValid()` that returns `implementation_delegate ? implementation_delegate( type ) : true` more straightforward
</blockquote></details>

Are you talking about the `CheckImplementationDelegate` function?


<!-- Sent from Reviewable.io -->

rascar-capac reviewed (changes requested):
> Reviewed 5 of 5 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110)* status: all files reviewed, 4 unresolved discussions (waiting on @fishingcactus-nathan-istace, @lucas-baran, @MaxDaniels, @PetitT, @Skeloson, and @ThibNach)

___
*[`Editor/ContextMenus/PropertyContextMenus/CreateAssetContextMenu.cs` line 124 at r4](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110#-NwivrluCLaw51oZ39xq:-NwxgFFA3PoGEsr2gUce:bas4184) ([raw file](https://github.com/FishingCactus/UnityCommonCode/blob/2b8645e3b248e77bc4eb423722d2f6fcb4bf2152/Editor/ContextMenus/PropertyContextMenus/CreateAssetContextMenu.cs#L124)):*
<details><summary><i>Previously, lucas-baran (Lucas Baran) wrote…</i></summary><blockquote>

Actually it could/should be a setting I think so that it follows the naming conventions of the project, tbh I won't be doing that now xD
</blockquote></details>

maybe add a TODO 
?
___
*[`Editor/DropDowns/Items/TypeDropdownItem.cs` line 9 at r4](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110#-Nwj-wlP5Ws2T-lCUXYK:-NwxhFiuB28dEB12cWXI:b-w2z0oo) ([raw file](https://github.com/FishingCactus/UnityCommonCode/blob/2b8645e3b248e77bc4eb423722d2f6fcb4bf2152/Editor/DropDowns/Items/TypeDropdownItem.cs#L9)):*
<details><summary><i>Previously, lucas-baran (Lucas Baran) wrote…</i></summary><blockquote>

What Nathan said, and we don't have a standards on this
</blockquote></details>

That would make no sense to have default initialization on fields, but not on auto-properties
And if we want to add that exception, it should be in the standard
___
*[`Editor/Utilities/InspectorUtilities.cs` line 62 at r4](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110#-Nwj6Ouf2nQJbXfW7Yj_:-Nwxhgto6qTuLjIBRhJY:bejjxf4) ([raw file](https://github.com/FishingCactus/UnityCommonCode/blob/2b8645e3b248e77bc4eb423722d2f6fcb4bf2152/Editor/Utilities/InspectorUtilities.cs#L62)):*
<details><summary><i>Previously, lucas-baran (Lucas Baran) wrote…</i></summary><blockquote>

Yeeah it's the same anyways right?
</blockquote></details>

Less readable imo but thats fine
___
*[`Runtime/Helpers/ReflectionUtilities.cs` line 34 at r4](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110#-Nwj59aK7iPsBT2cWo9C:-NwxiNpqCy2Ovl74v_0M:b-x1eaae) ([raw file](https://github.com/FishingCactus/UnityCommonCode/blob/2b8645e3b248e77bc4eb423722d2f6fcb4bf2152/Runtime/Helpers/ReflectionUtilities.cs#L34)):*
<details><summary><i>Previously, lucas-baran (Lucas Baran) wrote…</i></summary><blockquote>

Are you talking about the `CheckImplementationDelegate` function?
</blockquote></details>

It would get rid of that and also of `TrueImplementationDelegate` which I find a bit convoluted


<!-- Sent from Reviewable.io -->

MaxDaniels reviewed (approved):
> Reviewed 3 of 4 files at r2, 2 of 3 files at r4, 5 of 5 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110)* status: all files reviewed, 4 unresolved discussions (waiting on @fishingcactus-nathan-istace, @lucas-baran, @PetitT, @Skeloson, and @ThibNach)



<!-- Sent from Reviewable.io -->

fishingcactus-nathan-istace reviewed (approved):
> Reviewed 5 of 5 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110)* status: all files reviewed, 4 unresolved discussions (waiting on @lucas-baran, @PetitT, @Skeloson, and @ThibNach)



<!-- Sent from Reviewable.io -->

lucas-baran committed:
InspectorUtilities.GetPropertyType: use while loop instead of for

lucas-baran committed:
CreateAssetContextMenu: add TODO comment

lucas-baran committed:
ReflectionUtilities: add ImplementationIsValid to check valid implementations

lucas-baran committed:
ReflectionUtilities: add CanCreateInstanceOfImplementation to check implementations that can be instanced

lucas-baran committed:
ReflectionUtilities: remove CheckImplementationDelegate and TrueImplementationDelegate

lucas-baran reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110)* status: all files reviewed, 3 unresolved discussions (waiting on @PetitT, @rascar-capac, @Skeloson, and @ThibNach)

___
*[`Editor/ContextMenus/PropertyContextMenus/CreateAssetContextMenu.cs` line 124 at r4](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110#-NwivrluCLaw51oZ39xq:-NwyL0E32rvXDefZ4U7L:b-896fix) ([raw file](https://github.com/FishingCactus/UnityCommonCode/blob/2b8645e3b248e77bc4eb423722d2f6fcb4bf2152/Editor/ContextMenus/PropertyContextMenus/CreateAssetContextMenu.cs#L124)):*
<details><summary><i>Previously, rascar-capac (Théo Constant) wrote…</i></summary><blockquote>

maybe add a TODO 
?
</blockquote></details>

Done.
___
*[`Editor/DropDowns/Items/TypeDropdownItem.cs` line 9 at r4](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110#-Nwj-wlP5Ws2T-lCUXYK:-NwyL7hC5WA3Wws6zYgL:bfu03v0) ([raw file](https://github.com/FishingCactus/UnityCommonCode/blob/2b8645e3b248e77bc4eb423722d2f6fcb4bf2152/Editor/DropDowns/Items/TypeDropdownItem.cs#L9)):*
<details><summary><i>Previously, rascar-capac (Théo Constant) wrote…</i></summary><blockquote>

That would make no sense to have default initialization on fields, but not on auto-properties
And if we want to add that exception, it should be in the standard
</blockquote></details>

Why would that not make no sense? A property, even if it's an auto-property, is not a field thus doesn't have to follow the coding standards for fields. It's more like if we want to add a rule for default initialization of auto-properties, it should be in the standard, not the other way around.
___
*[`Editor/Utilities/InspectorUtilities.cs` line 62 at r4](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110#-Nwj6Ouf2nQJbXfW7Yj_:-NwyJgye8VbjshbIodY_:b-896fix) ([raw file](https://github.com/FishingCactus/UnityCommonCode/blob/2b8645e3b248e77bc4eb423722d2f6fcb4bf2152/Editor/Utilities/InspectorUtilities.cs#L62)):*
<details><summary><i>Previously, rascar-capac (Théo Constant) wrote…</i></summary><blockquote>

Less readable imo but thats fine
</blockquote></details>

Done.
___
*[`Runtime/Helpers/ReflectionUtilities.cs` line 45 at r1](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110#-NwEQ02k2qXVxy_0GI7j:-NwyN1BoC7YL5rjqer55:b8nzlt1) ([raw file](https://github.com/FishingCactus/UnityCommonCode/blob/ef4587e27035448c31d9ce16cc8eb88262b4eb53/Runtime/Helpers/ReflectionUtilities.cs#L45)):*
<details><summary><i>Previously, lucas-baran (Lucas Baran) wrote…</i></summary><blockquote>

So I tried to change it and I remember why I did it this way no, I'm using the function in the `GetImplementations` function which is a `Func<Type, bool>` but I also need the `implementation_delegate` so it allows me to capture it inside the local function which wouldn't be possible if it weren't local.
I could as well define a function with the conditions required for creating instances and use a lambda to combine `CanActivateImplementation` and `implementation_delegate` if you prefer
</blockquote></details>

In the end I still removed the local function and added a lambda, following the comment from Théo (https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110#-Nwj59aK7iPsBT2cWo9C). Makes it a bit better
___
*[`Runtime/Helpers/ReflectionUtilities.cs` line 34 at r4](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110#-Nwj59aK7iPsBT2cWo9C:-NwyN-L0AQyx9G76JElQ:b-896fix) ([raw file](https://github.com/FishingCactus/UnityCommonCode/blob/2b8645e3b248e77bc4eb423722d2f6fcb4bf2152/Runtime/Helpers/ReflectionUtilities.cs#L34)):*
<details><summary><i>Previously, rascar-capac (Théo Constant) wrote…</i></summary><blockquote>

It would get rid of that and also of `TrueImplementationDelegate` which I find a bit convoluted
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rascar-capac reviewed (approved):
> Reviewed 3 of 3 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110)* status: all files reviewed, 1 unresolved discussion (waiting on @PetitT, @Skeloson, and @ThibNach)

___
*[`Editor/DropDowns/Items/TypeDropdownItem.cs` line 9 at r4](https://reviewable.io/reviews/FishingCactus/UnityCommonCode/110#-Nwj-wlP5Ws2T-lCUXYK:-NwySBxU-_Qbtq1J8T3r:bqlmfvq) ([raw file](https://github.com/FishingCactus/UnityCommonCode/blob/2b8645e3b248e77bc4eb423722d2f6fcb4bf2152/Editor/DropDowns/Items/TypeDropdownItem.cs#L9)):*
<details><summary><i>Previously, lucas-baran (Lucas Baran) wrote…</i></summary><blockquote>

Why would that not make no sense? A property, even if it's an auto-property, is not a field thus doesn't have to follow the coding standards for fields. It's more like if we want to add a rule for default initialization of auto-properties, it should be in the standard, not the other way around.
</blockquote></details>

To be honest I find that convention a bit useless, but I guess that the reason we have it is to explicitely show what value a field has at declaration. If that's the case, I don’t see why we wouldn't do it as well for the properties' backing fields, since we use it the same way as a field in the private scope. We should show their value at declaration as well. Anyway, I'll make sure to discuss it when we create the auto-formatting tool :P


<!-- Sent from Reviewable.io -->
